ddtrace/debugging/_probe/registry.py

```diff
@@ -0,0 +1,185 @@
+            self._remove_pending(probe)
+
+            self.logger.installed(probe)
+
+    def set_exc_info(self, probe, exc_info):
+        # type: (Probe, ExcInfoType) -> None
+        """Set the installed flag for a probe."""
+        with self._lock:
+            self[probe.probe_id].set_exc_info(exc_info)
+            self.logger.error(probe, exc_info=exc_info)
+
+    def set_error(self, probe, message):
+        # type: (Probe, str) -> None
+        """Set the error message for a probe."""
+        with self._lock:
+            self[probe.probe_id].set_message(message)
+            self.logger.error(probe, message)
+
+    def log_probes_status(self):
+        # type: () -> None
+        """Log the status of all the probes using the status logger."""
+        with self._lock:
+            for entry in self.values():
+                if entry.installed:
+                    self.logger.installed(entry.probe)
+                elif entry.exc_info:
+                    self.logger.error(entry.probe, exc_info=entry.exc_info)
+                elif entry.message:
+                    self.logger.error(entry.probe, message=entry.message)
+                else:
+                    self.logger.received(entry.probe)
+
+    def _remove_pending(self, probe):
+        # type: (Probe) -> None
+        location = _get_probe_location(probe)
+
+        # Pending probes must have valid location information
+        assert location is not None, probe
+
+        pending_probes = self._pending[location]
+        try:
+            # DEV: Note that this is O(n), which is fine with a conservative
+            # number of probes.
+            pending_probes.remove(probe)
+        except ValueError:
+            # The probe wasn't pending
+            pass
+        if not pending_probes:
+            del self._pending[location]
+
+    def unregister(self, *probes):
+        # type: (Probe) -> List[Probe]
+        """Unregister a collection of probes.
+
+        This also ensures that any pending probes are removed if they haven't
+        been processed yet.
+        """
+        unregistered_probes = []
+        with self._lock:
+            for probe in probes:
+                try:
+                    entry = self.pop(probe.probe_id)
+                except KeyError:
+                    # We don't seem to have the probe
+                    logger.warn("Tried to unregister unregistered probe %s", probe.probe_id)
+                else:
+                    probe = entry.probe
+                    self._remove_pending(probe)
+                    unregistered_probes.append(probe)
+        return unregistered_probes
+
+    def get_pending(self, location):
+        # type: (str) -> List[Probe]
+        """Get the currently pending probes by location."""
+        return self._pending[location]
+
+    def __contains__(self, probe):
+        # type: (object) -> bool
+        """Check if a probe is in the registry."""
+        assert isinstance(probe, Probe)
```

Comment:
I'm guessing this should never happen, but if it does, should we catch the assertion error?

Reply:
This is an internal API, so we should really never fail this assert, but the traceback should be enough to investigate if necessary I think